cpp/src/arrow/scalar.cc

```diff
@@ -199,25 +199,30 @@ Result<std::shared_ptr<Scalar>> DictionaryScalar::GetEncodedValue() const {
     return MakeNullScalar(dict_type.value_type());
   }
 
+  const void* index =
+      checked_cast<const internal::PrimitiveScalarBase&>(*value.index).data();
```

Comment:
Why not cast to the target numeric scalar directly instead of making of virtual function call to `data()`?